Revisit request authentication (#305)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

# Message Transport

Communications between DAP participants are carried over HTTPS {{!RFC2818}}.
HTTPS provides server authentication and confidentiality. Use of HTTPS is

> REQUIRED.

Author: I think there's good arguments against relying on a transport-level concept of identity to authenticate protocol-level concepts, but [this is good enough for ACME](https://www.rfc-editor.org/rfc/rfc8555.html#section-6) so perhaps we can lean on that precedent.

Reviewer: Ultimately this question would be answered by a proof of (in)security in a formal model. I'm perfectly happy to make this a requirement for now.

One nit: Does RFC2818 cover HTTP/3? I'm wondering if @martinthomson can help us spell this requirement :)

Reviewer 2: You can start to refer to RFC 9110 for that stuff now.  That definitely covers HTTP/3.

(Technically, RFC2818 did cover HTTP/3, but only circuitously.  Also, RFC 2818 is a little old in a few important ways.)

# Message Transport

Communications between DAP participants are carried over HTTPS {{!RFC2818}}.
HTTPS provides server authentication and confidentiality. Use of HTTPS is
REQUIRED.

> ## Request Authentication {#request-authentication}

Reviewer: ?

```suggestion
## HTTP Request Authentication {#request-authentication}
```

DAP is made up of several sub-protocols in which different subsets of the
protocol's participants interact with each other.

In those cases where a channel between two participants is tunneled through

> another protocol participant, DAP mandates the use of authenticated encryption

Reviewer: Bump: We're not using HPKE for authenticated encryption here. The properties we need are confidentiality and non-malleability.
```suggestion
another protocol participant, DAP mandates the use of public-key encryption
```

using {{!HPKE=RFC9180}} to ensure that only the intended recipient can see a
message in the clear.

> In other cases, DAP requires client authentication. Any authentication scheme

Reviewer: We should try to avoid overloading "Client" wherever possible.

```suggestion
In other cases, DAP requires sender authentication. Any authentication scheme
```

that is composable with HTTP is allowed. For example, {{!OAuth2=RFC6749}}
credentials are presented in an Authorization HTTP header, which can be added to
any DAP protocol message. This allows organizations deploying DAP to use

> existing well-known HTTP authentication mechanisms that they already supports.

Reviewer 3: ```suggestion
existing well-known HTTP authentication mechanisms that they already support.
```

(typo)

Discovering what authentication mechanisms are supported by a DAP participant

> is outside of this document's scope.

Reviewer: It's worth mentioning mutual TLS here as a viable option.

The client MUST abort if any of the following happen for any HPKE config
request:

* the client and aggregator failed to establish a secure,

>   aggregator-authenticated channel;

Author: I deleted this because it's implicit by virtue of using HTTPS: you won't even be able to establish a connection if you can't verify server identity, so we don't need more protocol/application layer language to repeat features offered by the transport.

"extension_type" indicates the type of extension, and "extension_data" contains
information specific to the extension.

### Upload Message Security

The contents of each input share must be kept confidential from everyone but the

> client and the aggregator it is being sent to and clients must be able to

Reviewer: ```suggestion
client and the aggregator it is being sent to. In addition, clients must be able to
```

authenticate the aggregator they upload to.

HTTPS provides confidentiality between the client and the leader, but this is
not sufficient since the helper's report shares are relayed through the leader.
Confidentiality of report shares is achieved by encrypting each report share to
a public key held by the respective aggregator. Clients fetch the public keys
from each aggregator over HTTPS, allowing them to authenticate the server.

Aggregators MAY require clients to authenticate when uploading reports. This is
an effective mitigation against Sybil attacks in deployments where it is
practical for each client to have an identity provisioned (e.g., a user logged
into an online service or a hardware device programmed with an identity). If it
is used, client authentication MUST use a scheme that meets the requirements in
{{request-authentication}}.

In some deployments, it will not be practical to require clients to authenticate
(e.g., a widely distributed application that does not require its users to login

> to any service), so client authentication is not mandatory in DAP. In the
> absence of client authentication, implementations SHOULD deploy some mitigation
> against Sybil attacks.

Reviewer: As it stands, this problem is unsolved. I don't think it's fair to nudge implementors until we have a better sense of what alternatives are possible. It might be worth pointing to an issue regarding Sybil attacks (we have at least one somewhere).

```suggestion
to any service), so client authentication is not mandatory in DAP.
```

the collector.

Collectors MUST authenticate their requests to leaders using a scheme that meets
the requirements in {{request-authentication}}.

[[OPEN ISSUE: collector public key is currently in the task parameters, but this
will have to change #102]]

The collector and helper never directly communicate with each other, but the
helper does transmit an aggregate share to the collector through the leader, as
detailed in {{collect-aggregate}}. The aggregate share must be confidential from
everyone but the helper and the collector.

Confidentiality is achieved by having the helper encrypt its aggregate share to
a public key held by the collector.

There is no authentication between the collector and the helper. This allows the
leader to:

* Send collect parameters to the helper that do not reflect the parameters
  chosen by the collector
* Discard the aggregate share computed by the helper and then fabricate
  aggregate shares that combine into an arbitrary aggregate result

These are both attacks on soundness, which we already assume to hold only if
both aggregators are honest, which puts these malicious-leader attacks out of

> scope.

Reviewer: Point to security considerations here.

Author: I put in a reference to {{sec-considerations}} and use the word "correctness", which is currently the term that Security Considerations uses. That's wrong, but it'll get cleaned up in #306.

It also occurs to me that maybe everything in these "{Upload, Aggregate, Collect} Message Security" sections that isn't a MUST or a MAY should be punted to security considerations, since it's discussion, context and justification rather than prescription.

Reviewer: I think it's OK to point to security considerations as needed in the normative text, especially when it clarifies certain design choices to the reader. That said, mnimizing where possible is a good idea. In particular I'm not sure I would call these sections ".* Message Security".

[[OPEN ISSUE: Should we have authentication in either direction between the
helper and the collector? #155]]
[[OPEN ISSUE: DAP and/or VDAF should spell out what "soundness" and "privacy"

> mean]]

Reviewer: VDAF does already. Does DAP not as well?

Author: Oh, whoops, I was searching for "soundness" in VDAF, but it discusses "robustness". And in security considerations we discuss "privacy" and "correctness", but the doc has references to "soundness" and "robustness", both in protocol text and the odd OPEN ISSUE block. I filed #306 to clean up this language in DAP and align it with VDAF.
